src/Dialogs/MainWindow.vala

```diff
@@ -125,22 +125,8 @@ public class Tuba.Dialogs.MainWindow: Adw.ApplicationWindow, Saveable {
 	}
 
 	public Views.Base open_view (Views.Base view) {
-		if (
-			leaflet.visible_child == view
-			|| (
-				last_view != null
-				&& last_view.label == view.label
-				&& !view.allow_nesting
-			)
-		) return view;
-
-		if (last_view != null && !last_view.is_main && view.is_sidebar_item) {
-			leaflet.insert_child_after (view, main_base);
-		} else {
-			leaflet.append (view);
-		}
-
-		leaflet.visible_child = view;
```

Comment:
This part exists for two reasons:
1. Avoiding opening the current view again (example: if you are on "Lists" and click "Lists" again it won't open it again)
2. If you are in a sidebar view and select another it removes all other pages and then adds it (or rather View -> Home -> Another View) (removal happens elsewhere), this is to avoid circular navigation (Home -> Lists -> Hashtags -> Lists becomes Home -> Lists). Oh and for performance reasons (many of the views will keep an active websocket connection as long as they are not disposed). This has its cons but it's probably a discussion to be held in a different issue

Reply:
I see, thanks! I'll try to restore this logic then

Reply:
Pushed an update that tries to use `navigation_view.replace ()` for sidebar item pages. This is still not perfect, because now the Home row in the sidebar doesn't seem to have any effect at all. Also possibly this should be doing `navigation_view.replace ({ home_page, page });` and not just `{ page }`, to enable going back home from other sidebar views; and then also if we're already at `home`, we still should be doing `push ()`, not `replace ()`, so we get the animation (& building the intuition that it's possible to go back).

Please tell me what you think!

Reply:
I'm not that fond of replacing the whole stack every time for the reasons you mentioned, how about this:

```diff
diff --git a/src/Dialogs/MainWindow.vala b/src/Dialogs/MainWindow.vala
index fabce38..ca31ca1 100644
--- a/src/Dialogs/MainWindow.vala
+++ b/src/Dialogs/MainWindow.vala
@@ -19,7 +19,6 @@ public class Tuba.Dialogs.MainWindow: Adw.ApplicationWindow, Saveable {
 		var gtk_settings = Gtk.Settings.get_default ();
 	}
 
-	private Views.Base main_base;
 	public MainWindow (Adw.Application app) {
 		Object (
 			application: app,
@@ -28,8 +27,7 @@ public class Tuba.Dialogs.MainWindow: Adw.ApplicationWindow, Saveable {
 			resizable: true
 		);
 		sidebar.set_sidebar_selected_item (0);
-		main_base = new Views.Main ();
-		open_view (main_base);
+		open_view (new Views.Main (), "main");
 
 		if (Build.PROFILE == "development") {
 			this.add_css_class ("devel");
@@ -124,13 +122,29 @@ public class Tuba.Dialogs.MainWindow: Adw.ApplicationWindow, Saveable {
 		}
 	}
 
-	public Views.Base open_view (Views.Base view) {
-		var page = new Adw.NavigationPage (view, view.label);
-		if (view.is_sidebar_item) {
-			navigation_view.replace ({ page });
+	public Views.Base open_view (Views.Base view, string? tag = null) {
+		if (
+			navigation_view?.visible_page?.child == view
+			|| (
+				last_view != null
+				&& last_view.label == view.label
+				&& !view.allow_nesting
+			)
+		) return view;
+
+		Adw.NavigationPage page;
+		if (tag == null) {
+			page = new Adw.NavigationPage (view, view.label);
 		} else {
-			navigation_view.push (page);
+			page = new Adw.NavigationPage.with_tag (view, view.label, tag);
 		}
+
+		if (last_view != null && !last_view.is_main && view.is_sidebar_item) {
+			navigation_view.pop_to_tag ("main");
+		}
+
+		navigation_view.push (page);
+
 		return view;
 	}
 

```

It goes back to the page tagged as "main" and then pushes the new page

Reply:
Pushed an implementation, please try it out.

Reply:
Looks good!